lib/src/css_mirroring/rulesets_processor.dart

```diff
@@ -0,0 +1,82 @@
+part of scissors.src.css_mirroring.bidi_css_generator;
+
+abstract class NodeProcessor {
+  void process(EditConfiguration e, TopLevelsPair t, int i);
+}
+
+class DeclarationProcessor extends NodeProcessor {
+  final _PendingRemovals removableRuleSets;
+  final int parentEnd;
+
+  DeclarationProcessor(this.removableRuleSets, this.parentEnd);
+
+  @override
+  void process(EditConfiguration _editConfig, TopLevelsPair _topLevelsPair,
+      int _iTopLevel) {
+    final removals = new _PendingRemovals(_editConfig.transaction);
+    final List<TreeNode> usedTopLevels =
+        _getUsedTopLevels(_editConfig.mode, _topLevelsPair);
+
+    _storeRemovableDeclarations(
+        _editConfig, _topLevelsPair, _iTopLevel, removals);
+
+    if (_isRuleRemovable(
+        removals, _topLevelsPair._originalTopLevels[_iTopLevel])) {
+      removableRuleSets.addRemoval(_getNodeStart(usedTopLevels[_iTopLevel]),
+          _getRuleSetEnd(usedTopLevels, _iTopLevel, parentEnd));
+    } else {
+      removals.commitRemovals();
+
+      /// Add direction attribute to RuleId for direction-specific RuleSet.
+      if (_editConfig.mode != _RetentionMode.keepBidiNeutral) {
+        _prependDirectionToRuleSet(_editConfig, usedTopLevels[_iTopLevel]);
+      }
+    }
+  }
+
+  /// Stores start and end locations of removable declarations in a ruleset
+  /// based upon the retention mode.
+  void _storeRemovableDeclarations(
```

Comment:
This only stores the declarations to be removed and does not actually remove them. So shouldn't it be kept like this?